zephyr/scripts/patch_build_id.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python3
+
+import sys
+import os
+import hashlib
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import Symbol
+from elftools.elf.constants import SH_FLAGS
+from intelhex import IntelHex
+
+
+SPOTFLOW_BINDESC_BUILD_ID_SYMBOL_NAME = "bindesc_entry_spotflow_build_id"
+
+# Binary descriptor ID is 0x5f0, type is bytes, and length is 20
+BUILD_ID_HEADER_VALUE = b"\xf0\x25\x14\x00"
+
+BUILD_ID_HEADER_SIZE = len(BUILD_ID_HEADER_VALUE)
+BUILD_ID_VALUE_SIZE = 20
+
+
+def generate_and_patch_build_id(elf_filepath: str, other_filepaths: list[str]):
+    with open(elf_filepath, "rb+") as elf_stream:
+        elffile = ELFFile(elf_stream)
+
+        bindesc_build_id_symbol = find_bindesc_build_id_symbol(elffile)
+
+        build_id = generate_build_id(elffile, bindesc_build_id_symbol)
+
+        patch_build_id_parsed_elf(
+            elffile, elf_filepath, bindesc_build_id_symbol, build_id
+        )
+
+        for filepath in other_filepaths:
+            if not os.path.exists(filepath):
+                # It is easier to check it here than in CMake
+                continue
+
+            if (
+                filepath.endswith(".elf")
+                or filepath.endswith(".exe")
+                or filepath.endswith(".strip")
+            ):
+                patch_build_id_elf(filepath, bindesc_build_id_symbol, build_id)
+            elif filepath.endswith(".hex"):
+                patch_build_id_hex(filepath, bindesc_build_id_symbol, build_id)
+            elif filepath.endswith(".bin"):
+                patch_build_id_bin(filepath, elffile, bindesc_build_id_symbol, build_id)
+            else:
+                raise Exception(f"Unsupported file type: {filepath}")
+
+
+def find_bindesc_build_id_symbol(elffile: ELFFile) -> Symbol:
+    symbol = None
+
+    symbol_table = elffile.get_section_by_name(".symtab")
+    if symbol_table is None:
+        symbol_table = elffile.get_section_by_name(".dynsym")
+
+    if symbol_table is not None:
+        symbol = symbol_table.get_symbol_by_name(SPOTFLOW_BINDESC_BUILD_ID_SYMBOL_NAME)
+
+    if symbol is None:
+        raise Exception(f"Symbol '{SPOTFLOW_BINDESC_BUILD_ID_SYMBOL_NAME}' not found")
+
+    if len(symbol) > 1:
+        raise Exception(
+            f"Symbol '{SPOTFLOW_BINDESC_BUILD_ID_SYMBOL_NAME}' found multiple times"
+        )
+
+    return symbol[0]
+
+
+def generate_build_id(elffile: ELFFile, bindesc_build_id_symbol: Symbol):
+    hash_builder = hashlib.sha1()
+
+    for section in elffile.iter_sections():
+        if section["sh_flags"] & SH_FLAGS.SHF_ALLOC == 0:
```

Comment:
Add comment explaining this filter (to only load sections needed at runtime, right)?